src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,122 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
     }
   }
 
+  // Convert a + a + ... + a into a*n
+  Node* serial_additions = convert_serial_additions(phase, bt);
+  if (serial_additions != nullptr) {
+    return serial_additions;
+  }
+
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Try to convert a serial of additions into a single multiplication. Also convert `(a * CON) + a` to `(CON + 1) * a` as
+// a side effect. On success, a new MulNode is returned.
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, BasicType bt) {
+  // We need to make sure that the current AddNode is not part of a MulNode that has already been optimized to a
+  // power-of-2 addition (e.g., 3 * a => (a << 2) + a). Without this check, GVN would keep trying to optimize the same
+  // node and can't progress. For example, 3 * a => (a << 2) + a => 3 * a => (a << 2) + a => ...
+  if (find_power_of_two_addition_pattern(this, bt).valid) {
+    return nullptr;
+  }
+
+  Node* in1 = in(1);
+  Node* in2 = in(2);
+
+  // While multiplications can be potentially optimized to power-of-2 subtractions (e.g., a * 7 => (a << 3) - a),
+  // (x - y) + y => x is already handled by the Identity() methods. So, we don't need to check for that pattern here.
+  Multiplication mul;
+  if (
+      ((mul = find_simple_addition_pattern(in1, bt)).valid && mul.variable == in2) ||
+      ((mul = find_simple_lshift_pattern(in1, bt)).valid && mul.variable == in2) ||
+      ((mul = find_simple_multiplication_pattern(in1, bt)).valid && mul.variable == in2) ||
+      ((mul = find_power_of_two_addition_pattern(in1, bt)).valid && mul.variable == in2)
+      ) {
```

Comment:
I find this quite difficult to read. And it looks repetitive too. Maybe you can refactor it?
Also, it would be nice to have comments with the patterns here, to see which one covers what case, so that we have a nice overview.

Reply:
Refactored. Added comments with example. It's still some-what repetitive and definitely much clearer.

https://github.com/openjdk/jdk/blob/eefe3a35effc469f3527d4825c9cc4f2a72742dc/src/hotspot/share/opto/addnode.cpp#L421-L444